### PR TITLE
Style RC parameter

### PR DESCRIPTION
**Pull request by Mrngilles**
Added a style RC parameter to be able to import styles from stylesheet or matplotlibrc.
You can call a parent stylesheet using the following syntax in your matplotlibrc or a stylesheet: 
'style : chosen_style' if the style is in the library (built-in styles or a style put in your user config path
'style : path/to/style' to get a style from anywhere.

I also added a test to make sure we get the good rcParameters when calling parent styles. For that, I added sample stylesheets in the `sample_data parameter`, that form the same tree as the one seen in the `get_style_parameters` function. It should cover every possible override possible.

_Priority order remark :_ When calling styles this way, the first ones will have the highest priority.

Todo:
- Breakdown in smaller functions to follow the flow more easily

### CODE REVIEW

**Comment by Mrngilles** [comment]
I get errors from travis that my stylesheets cannot be found... I am not sure how I could get it to work. Also, it is now in the sample_data folder, which I think is not the best location for data that will be used only for tests.
Do you have any idea how I could solve this?

**Comment by tonysyu** [comment]
Thanks for breaking this out into its own PR!

A general comment about the tests and sample style files: I think all of the style files are unnecessary. Having tests that depend on values that are in a separate file makes the logic of the tests hard to follow. Instead, I would suggest reusing the `temp_style` context manager to create a couple of style files. (BTW, I noticed that I neglected to properly connect the `settings` argument of `temp_style`; there should have been an `if settings is None` before using the default.) That way, the test inputs and outputs are clear.

File: lib/matplotlib/rcsetup.py [hunk 1]
```diff
@@ -501,6 +501,8 @@ def __call__(self, s):
 
 # a map from key -> value, converter
 defaultParams = {
+    'style': ['', validate_stringlist],
```
Comment (tonysyu) [inline lib/matplotlib/rcsetup.py]: As mentioned in the original PR, I don't think it's important to support a list of parent stylesheets---at least not initially. I could be convinced otherwise, but I think it adds complication to the logic.

File: lib/matplotlib/style/__init__.py [hunk 2]
```diff
@@ -1,3 +1,6 @@
 from __future__ import absolute_import
 
 from .core import use, context, available, library, reload_library
+from matplotlib import rcParams
+if rcParams['style']:
+    use(rcParams['style'])
```
Comment (tonysyu) [inline lib/matplotlib/style/__init__.py]: I don't think this is the appropriate place for this call. IIUC, the default style isn't used unless the `style` module is imported.

File: lib/matplotlib/style/core.py [hunk 3]
```diff
@@ -66,6 +67,12 @@ def use(style):
     else:
         styles = style
 
+    tmp_styles = []
+    for style in styles:
+        style = get_parents_styles(style)
+        tmp_styles.extend(style)
+    styles = tmp_styles[:]
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: I think this should be refactored so this diff is a one-liner; e.g. `styles = insert_parent_styles(styles)` ("insert" sort of implies an in-place operation, so a different verb might be appropriate)

File: lib/matplotlib/style/core.py [hunk 4]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+       |- c1
+       |- c2
+    This means that `main` calls `a`, `b`, and `c`
+    Similar pattern for a, b and c.
+    When calling `matplotlib.style.use(['main']), we will call all the
+    stylesheets recursively add apply all styles. In this example, it would
+    return the following list
+    ['c2', 'c1', 'c', 'b2', 'b1', 'b', 'a2', 'a1', 'a', 'main']
+    """
+    styles = [styles]
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: Something seems wrong here. If I'm reading this correctly, the loop below just runs a single time.

File: lib/matplotlib/style/core.py [hunk 5]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+    has_parent = True
+    full_styles = []
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: It looks like you're using this to detect cycles. I would call this `visited_styles`.

File: lib/matplotlib/style/core.py [hunk 6]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+            if _is_dict_like(style):
+                continue
+
+            style_dic = get_style_parameters(style)
+
+            if style_dic.get('style', None):
+                has_parent = True
+                parent = style_dic['style']
+                temp_styles.insert(temp_styles.index(style)+1, parent)
+                temp_styles = list(_list_flatten(temp_styles))
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: I think the bulk of this loop would be simplified if only a single parent style is allowed.

File: lib/matplotlib/style/core.py [hunk 7]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+
+    while has_parent:
+        has_parent = False
+        for style in styles:
+            if style in full_styles:
+                continue
+            full_styles.append(style)
+
+            if _is_dict_like(style):
+                continue
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: Is there any reason you decided not to support parent styles in a style-dict? I could either way.

File: lib/matplotlib/style/core.py [hunk 8]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+
+            style_dic = get_style_parameters(style)
+
+            if style_dic.get('style', None):
+                has_parent = True
+                parent = style_dic['style']
+                temp_styles.insert(temp_styles.index(style)+1, parent)
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: What if the parent defines a parent `style` of its own?

File: lib/matplotlib/style/core.py [hunk 9]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+    or an OrderedDict."""
+    if type(style) is dict:
+        return True
+    if sys.version_info[:2] >= (2, 7):
+        from collections import OrderedDict
+        if type(style) is OrderedDict:
+            return True
+
+
+def get_style_parameters(style):
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: This looks like it's duplicating a lot of the functionality in the `use` function.

File: lib/matplotlib/style/core.py [hunk 10]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+    """Get the parameter of the style, whether it is a file or a
+    dictionnary"""
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: This doesn't handle dictionaries, does it?

File: lib/matplotlib/style/core.py [hunk 11]
```diff
@@ -84,6 +91,101 @@ def use(style):
...
+                parent = style_dic['style']
+                temp_styles.insert(temp_styles.index(style)+1, parent)
+                temp_styles = list(_list_flatten(temp_styles))
+
+        styles = temp_styles[:]
+    styles = styles[::-1]
+    return styles
+
+
+def _is_dict_like(style):
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: Matplotlib may already have implementations of this (I'm no sure), but there are standard implementations. For example: http://stackoverflow.com/a/1278070/260303

File: lib/matplotlib/tests/test_style.py [hunk 12]
```diff
@@ -135,6 +152,23 @@ def test_context_with_badparam():
...
+            assert mpl.rcParams['lines.marker'] == 'o'
+            assert mpl.rcParams['lines.linestyle'] == '--'
+            assert mpl.rcParams['lines.markersize'] == 12
+            assert mpl.rcParams['axes.facecolor'] == 'green'
+            assert mpl.rcParams['axes.linewidth'] == 5.0
+            assert mpl.rcParams['axes.grid'] == True
+            assert mpl.rcParams['axes.labelweight'] == 'bold'
+            assert mpl.rcParams['axes.labelcolor'] == 'yellow'
+            assert mpl.rcParams['axes.color_cycle'] == [u'r', u'g', u'b',
+                                                            u'c', u'y']
```
Comment (tonysyu) [inline lib/matplotlib/tests/test_style.py]: As mentioned in my overall PR comment, it's difficult to actually verify this test because the input data is in a separate file (i.e. disconnected from the test code). Also, it's probably not necessary to test all of these values. For most tests, I think checking one value is sufficient.

File: lib/matplotlib/rcsetup.py (hunk 1 above)
Comment (Mrngilles) [inline lib/matplotlib/rcsetup.py]: Actually, getting a list of parent stylesheets is kind of my main idea. This way, you could have really small stylesheets setting only a part of the plot (*e.g.* only the font, or the lines width) that could be associated to create easily bigger stylesheets.

File: lib/matplotlib/style/__init__.py (hunk 2 above)
Comment (Mrngilles) [inline lib/matplotlib/style/__init__.py]: That is true, but I was not able to put it anywhere else... I tried to import the style from rc on matplotlib import, but it is not possible to do it from the mpl.__init__ as the style module uses some functions from the __init__.

File: lib/matplotlib/style/core.py (hunk 3 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: This was the plan, I  didn't have time to do it yet.

File: lib/matplotlib/style/core.py (hunk 7 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: Do you mean having parent style in a stylesheet in the dict form? If that is so, I don't really see the point as you can put the parameters in the stylesheet.
If you mean on the `use` call, dict are actually supported. as you may see, I append the style (dict or string) on `full_string` (that will change to `visited_styles`). This way, when returning all the styles to the `use` function, the dict will be processed.

File: lib/matplotlib/style/core.py (hunk 8 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: This is why I check if the currently analysed style has parent styles, and use the while loop. We will rescan the styles list until there are no parent styles left

File: lib/matplotlib/style/core.py (hunk 9 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: Indeed, I will use it in the `use` function.

File: lib/matplotlib/style/core.py (hunk 11 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: I will check, I think there might be

File: lib/matplotlib/style/core.py (hunk 10 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: Yes, I'll had it

File: lib/matplotlib/tests/test_style.py (hunk 12 above)
Comment (Mrngilles) [inline lib/matplotlib/tests/test_style.py]: Ok. Do you think that testing only one parent would be enough? By doing this, I wanted to try and have multiple calls to parent styles and parents of parents. Of course, I could do something simpler, using only 3 styles to try and have a style, its parent and the parent of its parent. I guess that testing this would be needed.
By testing all those values, I wanted to try and see if all the stylesheets are actually called and overrides the previous one. Again, I could also make it easier.

File: lib/matplotlib/style/core.py (hunk 11 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: I was not able to find any function to do this... If anyone does know such a function, please let me know.

File: lib/matplotlib/style/__init__.py (hunk 2 above)
Comment (Mrngilles) [inline lib/matplotlib/style/__init__.py]: After a small test, I confirmed something I already noticed before, which is that a style call from the matplotlibrc *will* be applied even without calling the `matplotlib.style` library. I guess that on init, matplotlib initializes also the style library, or giving the style parameter in the rc will call the style library... I'm not sure about which it is

File: lib/matplotlib/rcsetup.py (hunk 1 above)
Comment (tonysyu) [inline lib/matplotlib/rcsetup.py]: Personally, I see two broad categories for style sheets: Those focused on aesthetics and those focused on layout (for slides, journal papers, etc.). I think these should be combined at runtime with `style.use`. I don't see the need for more fine grained partitioning, but I imagine others would. I don't want to push too much, but I think a simpler initial implementation is a good idea.

File: lib/matplotlib/style/core.py [hunk 13]
```diff
@@ -66,22 +67,92 @@ def use(style):
...
+    temp_styles = styles[:]
+
+    while has_parent:
+        for style in styles:
+            if style in visited_styles:
+                continue
+            visited_styles.append(style)
+
+            if _is_dict_like(style):
+                style_dic = style
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: Rename: `style_dic` -> `style_dict`

File: lib/matplotlib/style/core.py [hunk 14]
```diff
@@ -66,22 +67,92 @@ def use(style):
...
+        mpl.rcParams.update(get_style_parameters(style))
+
+
+def insert_parent_styles(styles):
+    """Returns a list with all the styles and parents styles called"""
+    tmp_styles = []
+    for style in styles:
+        style = _get_parent_styles_from_str_or_dict(style)
+        tmp_styles.extend(style)
+    return tmp_styles[:]
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: Is this function equivalent to:
```python
_list_flatten([s for s in _get_parent_styles_from_str_or_dict(style)])
```

File: lib/matplotlib/style/core.py [hunk 15]
```diff
@@ -66,22 +67,92 @@ def use(style):
...
+        return True
+    if sys.version_info[:2] >= (2, 7):
+        from collections import OrderedDict
+        if type(style) is OrderedDict:
+            return True
+
+
+def get_style_parameters(style):
+    """Get the parameter rc of the style, whether it is a file or a
+    dictionnary"""
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: I don't mean to be repetitive, but you might have missed my earlier comment: This function doesn't handle dictionaries, so this docstring is misleading.

File: lib/matplotlib/style/core.py [hunk 16]
```diff
@@ -66,22 +67,92 @@ def use(style):
...
+            if style in visited_styles:
+                continue
+            visited_styles.append(style)
+
+            if _is_dict_like(style):
+                style_dic = style
+            else:
+                style_dic = get_style_parameters(style)
+
+            has_parent = 'style' in style_dic
```
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: I don't think this is sufficient to traverse a tree. If earlier nodes had a parent, but the last one doesn't, then those parents aren't ever visited, right?

File: lib/matplotlib/style/core.py (hunk 11 above)
Comment (tonysyu) [inline lib/matplotlib/style/core.py]: If there's not an implementation in Matplotlib, I think the one in the link above is a bit clearer.

File: lib/matplotlib/tests/test_style.py (hunk 12 above)
Comment (tonysyu) [inline lib/matplotlib/tests/test_style.py]: > Ok. Do you think that testing only one parent would be enough?

Unless you have some specific issue in mind, I think the base case (a stylesheet with no parent) and the inductive step (http://en.wikipedia.org/wiki/Mathematical_induction). There could be issues beyond this, but personally, I think its best to deal with them only if there's a specific issue that could present issues.

**Comment by tacaswell** [comment]
Re-milestoned this as 'proposed next point release'  which mean it will not be a blocker for the next point release.

This looks like it still has a good deal of work?

File: lib/matplotlib/style/core.py (hunk 16 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: In here, what I do is:
- Get children of the style to examine and put them in the `temp_styles`
- Then I replace the styles to be scanned by the `temp_styles`, so that I have all the brothers on a level.
- Then each of those styles is tested to be if it already has been scanned, and skipped unless it has not, meaning that even though earlier nodes are not scanned right away, they will be right after the child in the last generation of the current branch is scanned.

I am not sure this is quite clear, but I tried every scenario I could and this function always gave me the complete list of children (1 node with 3 children, 3 children each worked)

File: lib/matplotlib/style/core.py (hunk 14 above)
Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: Well, the result is the same, except in this case, we just need to add a list at the end of an other one, hence the use of `extend`. I wrote the `_list_flatten` function to be able to flatten the list after inserting objects in the middle of it, because having strings was giving me some trouble. It is not necessary to use it here, but I can change it if you think it would be clearer.

File: lib/matplotlib/tests/test_style.py (hunk 12 above)
Comment (Mrngilles) [inline lib/matplotlib/tests/test_style.py]: So are you saying I should remove the tests altogether? I'm not sure to understand

File: lib/matplotlib/rcsetup.py (hunk 1 above)
Comment (Mrngilles) [inline lib/matplotlib/rcsetup.py]: On the top on being able to create stylesheets by "cascading" them (it's a very small implementation of cascading...), you could have a style in your matplotlibrc so that you don't have to call the `style.use` function every time you create a new script.

File: lib/matplotlib/tests/test_style.py (hunk 12 above)
Comment (tonysyu) [inline lib/matplotlib/tests/test_style.py]: Sorry, I was a bit unclear. I just meant that it's probably only necessary to test for a single parent. I could also see the argument for testing a grandparent, but I don't think it's strictly necessary.The main point though, was that separating the test data in external files makes the tests difficult to verify. Also, a single test here is actually testing a number of different cases (inheritance from a parent, inheritance from a grandparent, multiple parents, multiple grandparents). It would be nice to separate these out into individual tests.

File: lib/matplotlib/rcsetup.py (hunk 1 above)
Comment (tonysyu) [inline lib/matplotlib/rcsetup.py]: I think you're right, actually: Combining multiple stylesheets in your default `matplotlibrc` file will probably be a common use case.

That said, I think the implementation might be a bit clearer with a different approach. Have you thought of writing this using recursive dictionary updates instead of looping over parent styles? I've played around with an alternate implementation if you're open to the idea.

Comment (Mrngilles) [inline lib/matplotlib/rcsetup.py]: Can you explain a bit more on detail about the recursive dictionary update? I'm willing to do the best implementation possible for this, so if this can be cleaner... 

Envoyé depuis mon appareil Samsung

-------- Message d'origine --------
De : Tony S Yu <notifications@github.com> 
Date : 22/06/2015  06:58  (GMT+01:00) 
À : matplotlib/matplotlib <matplotlib@noreply.github.com> 
Cc : Gilles Marin <gillesmarin1706@gmail.com> 
Objet : Re: [matplotlib] Style RC parameter (#4240) 

In lib/matplotlib/rcsetup.py:
> @@ -501,6 +501,8 @@ def __call__(self, s):
>  
>  # a map from key -> value, converter
>  defaultParams = {
> +    'style': ['', validate_stringlist],

I think you're right, actually: Combining multiple stylesheets in your default matplotlibrc file will probably be a common use case.

That said, I think the implementation might be a bit clearer with a different approach. Have you thought of writing this using recursive dictionary updates instead of looping over parent styles? I've played around with an alternate implementation if you're open to the idea.

Comment (tonysyu) [inline lib/matplotlib/rcsetup.py]: Here's a gist with a function that flattens a dictionary containing a special key pointing to "parent" dictionaries: https://gist.github.com/tonysyu/b02be7df210c4369afa6. That's a generic implementation with nothing specific to matplotlib styles. To use in this case you could write something like the following.

```python
PARENT_STYLES = 'styles'
flattened_style = flatten_style_dict({PARENT_STYLES: styles})

def flatten_style_dict(styles_dict):
    return flatten_inheritance_dict(styles_dict, PARENT_STYLES,
                                    expand_parent=_expand_parent)


def _expand_parent(parent_style):
    if cbook.is_string_like(parent_style):
        return get_style_parameters(parent_style)
    return parent_style
```

To actually integrate this with the `style` module, you'd want to remove the loop from `use` and just  call update on the `flattened_style` dictionary. Note that the current tests won't pass with this function because the order of styles in the `PARENT_STYLES` is flipped. If you flip the order of parent styles in your `mplstyle` files, the test will pass.

**Comment by WeatherGod** [comment]
@tacaswell , @Mrngilles , are there any outstanding issues here?

**Comment by Mrngilles** [comment]
There should not be. This should work, but could be improved. I will have to, and will as soon as possible.

TODO :
- Tests reduction
- Look in more details into the dict flattening as suggested by @tonysyu

**Comment by WeatherGod** [comment]
Given that we want to do a feature freeze (today, preferably), do you have
any objections to merging this now as-is (note, I have not personally
reviewed this PR)

**Comment by Mrngilles** [comment]
I have no objection, as this should be working. 

On July 13, 2015 3:50:00 PM GMT+02:00, Benjamin Root <notifications@github.com> wrote:
>Given that we want to do a feature freeze (today, preferably), do you
>have
>any objections to merging this now as-is (note, I have not personally
>reviewed this PR)
>
>On Sun, Jul 12, 2015 at 1:05 PM, Gilles Marin
><notifications@github.com>
>wrote:
>
>> There should not be. This should work, but could be improved. I will
>have
>> to, and will as soon as possible.
>>
>> TODO :
>>
>>    - Tests reduction
>>    - Look in more details into the dict flattening as suggested by
>>    @tonysyu <https://github.com/tonysyu>
>>
>> —
>> Reply to this email directly or view it on GitHub
>>
><https://github.com/matplotlib/matplotlib/pull/4240#issuecomment-120741237>
>> .
>>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/matplotlib/matplotlib/pull/4240#issuecomment-120934033

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by WeatherGod** [comment]
Ok, I'll look through it today and make sure it looks good to me (I will also need to update my classic-mode PR a bit, I think).

**Comment by tonysyu** [comment]
I believe this implementation reverses the order of precedence compared to style.use. I have other issues with the current state of the PR, but that is probably the most significant.

**Comment by tacaswell** [comment]
I would like to make sure Tony's concerns are addressed.

**Comment by Mrngilles** [comment]
Indeed. I will test it, and if the issue pointed out by Tony is present, it will be better to wait until I include the dictionary flattening method before merging 

On July 13, 2015 4:22:00 PM GMT+02:00, Thomas A Caswell <notifications@github.com> wrote:
>I would like to make sure Tony's concerns are addressed.
>
>On Mon, Jul 13, 2015 at 10:02 AM Tony S Yu <notifications@github.com>
>wrote:
>
>> I believe this implementation reverses the order of precedence
>compared to
>> style.use. I have other issues with the current state of the PR, but
>that
>> is probably the most significant.
>>
>> —
>> Reply to this email directly or view it on GitHub
>>
><https://github.com/matplotlib/matplotlib/pull/4240#issuecomment-120938052>
>> .
>>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/matplotlib/matplotlib/pull/4240#issuecomment-120944937

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by Mrngilles** [comment]
Except for Python 2.6, I'm not sure the errors are related to the PR...
As for 2.6, the `ERROR: Failure: AttributeError ('module' object has no attribute 'test_style')` is raised. Any idea where this could come from?

**Comment by tacaswell** [comment]
You are using some non 2.6 compatible code, probably a dictionary
comprehension (on phone so hard to check)

**Comment by Mrngilles** [comment]
That may be it, but the error does not point to any file I touched, only to `nose` files... Here is the whole error : 

```
ERROR: Failure: AttributeError ('module' object has no attribute 'test_style')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.6/site-packages/nose/loader.py", line 407, in loadTestsFromName
    module = resolve_name(addr.module)
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.6/site-packages/nose/util.py", line 322, in resolve_name
    obj = getattr(obj, part)
AttributeError: 'module' object has no attribute 'test_style'
```
I checked for non-compatible 2.6 code, but I didn't find any, although I may miss it.

**Comment by tacaswell** [comment]
That very helpful error message means 'nose could not import the module matplotlib.tests.test_style but silently snarfed the exception and are now telling you that the explicitly white listed modules to test do not exist'.

Are you using `assert_raises` as a context manager?

**Comment by tacaswell** [comment]
I kicked the 2.7 and 3.3 tests, they look like timeout issues we intermittently have with latex

**Comment by Mrngilles** [comment]
Indeed the tests for the dictionary flattening function use assert_raise as a context manager. Do you think changing that will make it work?

**Comment by tacaswell** [comment]
Rather than iterating with travis, it is probably easiest to install a 2.6 enviroment with conda and see what is failing on import.

**Comment by WeatherGod** [comment]
Yes, I got bitten by this once before. For some reason, assert_raise() is
not available as a context manager in python 2.6.

**Comment by Mrngilles** [comment]
Maybe that was a problem, but it was not the only one. `assert_is_not` and `assert_is_instance` were not in nose in python 2.6, so could not be imported.

**Comment by tacaswell** [comment]
This needs a re-base again.

**Comment by Mrngilles** [comment]
I am going to rebase this against `v1.5.x`. Should I also create a new PR against `v1.5.x` or can I leave it against `master`?

**Comment by tacaswell** [comment]
Leave it against master, we will cherry-pick the merge back to 1.5.x.

**Comment by Mrngilles** [comment]
Hi guys, 
Sorry for the delay. I finally rebased my PR and squashed everything.
I think it's clearer this way.
Hope it'll do

File: lib/matplotlib/style/core.py [hunk 17]
```diff
@@ -65,7 +66,7 @@ def _apply_style(d, warn=True):
     mpl.rcParams.update(_remove_blacklisted_style_params(d, warn=warn))
 
 
-def use(style):
+def use(styles):
```
Comment (tacaswell) [inline lib/matplotlib/style/core.py]: bit concerned about the API break on changing the arg name.

Comment (Mrngilles) [inline lib/matplotlib/style/core.py]: I agree, I did not think about this.
Then, shoud I just adapt to keep the same API?

Comment (tacaswell) [inline lib/matplotlib/style/core.py]: Yeah, just leave it as `style` and accept the some-what dodge plural issues.

**Comment by Mrngilles** [comment]
New rebase on top of master.

**Comment by Mrngilles** [comment]
Not sure why the tests failed, but doesn't seem related to my rebase.

File: lib/matplotlib/style/core.py [hunk 18]
```diff
@@ -91,26 +92,91 @@ def use(style):
...
-                msg = ("'%s' not found in the style library and input is "
-                       "not a valid URL or path. See `style.available` for "
-                       "list of available styles.")
-                raise IOError(msg % style)
+            parent_style = get_style_dict(parent_style)
+    return parent_style
+
+
+def flatten_inheritance_dict(child_dict, parent_key,
+                             expand_parent=lambda x: x):
```
Comment (tacaswell) [inline lib/matplotlib/style/core.py]: and why leave this with a default value like this?

**Comment by tacaswell** [comment]
That test has been flaky recently.

There is some weird public vs private function things (there is a public function which is used in a private function which is called from a public function with is called from a private function which is used in `use`?).  Not saying this is _wrong_, but it seems it could be simpler.

**Comment by jklymak** [comment]
This kind of got dropped.  I'm also not 100% sure about the reason why someone would want this as an rcparam, nor has there been a user base clamouring for this.  But please do re-open if I'm mistaken.
